### ISSUE TITLE
Some fonts not rendering in iOS 18

### ISSUE THREAD

**Issue by dskuza** [harfbuzz/harfbuzz#4891]
I'm currently running into an issue where certain fonts in iOS 18 are not rendering glyphs _at all_ when run on an iOS device, or simulator, running iOS 18. The same font works in iOS 17.x, so there might be something at the font-level Apple has changed. I'm not sure.

The font in question is `PingFang SC`, installed on both iOS and macOS.

I'm curious if this is something that is currently known, being addressed, or unknown. If this is known and there is a fix, please let me know so that I can try implementing it on my end. Googling the above font, it appears that it also has caused issues in other apps running macOS Sequoia (macOS 15).

## Font Locations

The font files can be obtained if you have Xcode(s) installed, and the iOS 17 and iOS 18 simulators by running something similar to 
`$ find /Library/Developer -type f -iname "PingFang*`

For myself, these were the following directories:

#### iOS 17 

`/Library/Developer/CoreSimulator/Volumes/iOS_21F79/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 17.5.simruntime/Contents/Resources/RuntimeRoot/System/Library/Fonts/LanguageSupport/PingFang.ttc`

#### iOS 18

`/Library/Developer/CoreSimulator/Volumes/iOS_22A3351/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 18.0.simruntime/Contents/Resources/RuntimeRoot/System/Library/PrivateFrameworks/FontServices.framework/CorePrivate/PingFangUI.ttc`

## hb-view

I ran the above files through `hb-view`, ensuring I was using the latest version. Running `hb-view --version` returned `10.0.1` The test glyph was `U+79C1`

![image](https://github.com/user-attachments/assets/1921c67d-6ccb-4daa-a422-751a5a1d0159)

(Ignore that iOS 16 and iOS 16 are mislabeled directory names)

Other apps such as `FontGoggles` are also unable to render the font correctly, but I don't know what that app uses under the hood. Just another anecdotal point that the iOS 18 font is tripping some libraries and apps up.

**Comment by khaledhosny** [comment]
See https://gitlab.freedesktop.org/freetype/freetype/-/issues/1281, the font has no usable glyph outlines data.

**Comment by dskuza** [comment]
Got it. From the sounds of it (incl. reading that issue), it seems unlikely that this'll be solved near-term. If you feel that's the case, feel free to close this issue.

**Comment by khaledhosny** [comment]
You should be able to draw such fonts using the new CoreText font functions.

**Comment by behdad** [comment]
> You should be able to draw such fonts using the new CoreText font functions.

The face also needs to be created using the new `hb_coretext_face_create_from_file_or_fail()` with the path of the system PingFang. 

You need to build HarfBuzz with coretext feature enabled.

On MacOS you can use:
```
$ hb-view --face-loader coretext --font-funcs coretext /System/Library/PrivateFrameworks/FontServices.framework/Versions/A/Resources/Reserved/PingFangUI.ttc -u 6868`
```

**Comment by yisibl** [comment]
It looks like Apple is using a new `hvgl` table.
![image](https://github.com/user-attachments/assets/4e89fc22-3316-4e6d-8b5c-b207cc378d66)

Someone wrote an article: 
- https://zhuanlan.zhihu.com/p/703335162
- https://bbs.pcbeta.com/viewthread-2016718-1-1.html

**Comment by dskuza** [comment]
I haven't tested on macOS, but on iOS 18, I'm seeing some interesting results.

On some devices, the original `hb_coretext_font_create` works; in some conditions it doesn't work _until an international keyboard has been added_, such as Japanese.

Using the new function, I'm not seeing it work on iOS 18(.1) on a fresh install using something similar to the following:

```
auto face = hb_coretext_face_create_from_file_or_fail(filepath, index);
if (face)
{
    auto font = hb_font_create(face);
    // ...use the font here
}
return nullptr;
```

Not that there's necessarily a response needed, here, just pointing out an interesting finding.

**Comment by behdad** [comment]
@dskuza Do you open the font at its system location?

**Comment by dskuza** [comment]
@behdad Yes, on iOS it's in a different location, though: `/System/Library/PrivateFrameworks/FontServices.framework/CorePrivate/PingFangUI.ttc`

**Comment by dskuza** [comment]
Great, thank you!

**Comment by hughkaun** [comment]
> [@behdad](https://github.com/behdad) Yes, on iOS (at least devices) it's in a different location, though: `/System/Library/PrivateFrameworks/FontServices.framework/CorePrivate/PingFangUI.ttc` sussed out from finding files within the simulator. I can verify a file at that location does exist, though, on device using iOS's `FileManager`
> 
> Edit: another side-note - quick desktop test against the simulator: if I run the above mentioned `hb-view` command against the file in the simulator directory, the character does render. However, if I copy the file into another directory, it's unable to be used and prints an error.

This works for me after add '/System/Library/PrivateFrameworks/FontServices.framework/CorePrivate'
